sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldHaveCorrectSignature.cs

```diff
@@ -0,0 +1,124 @@
+        private const string MakePublicMessage = "'public'";
+        private const string MakeNonAsyncMessage = "non-'async'";
+        private const string MakeNotGenericMessage = "non-generic";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> TrackedTestAttributes = ImmutableHashSet.Create(
+            KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_TestMethodAttribute,
+            KnownType.NUnit_Framework_TestAttribute,
+            KnownType.Xunit_FactAttribute);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    if (!c.IsTest())
+                    {
+                        return;
+                    }
+
+                    var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                    var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+                    if (classSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var allFaultyMethods = classSymbol.GetMembers()
+                        .OfType<IMethodSymbol>()
+                        .Where(method => method.GetAttributes()
+                            .Any(attribute => attribute.AttributeClass.IsAny(TrackedTestAttributes)))
+                        .Select(method => new
+                        {
+                            Location = method.Locations.First(),
+                            Message = ConcatMessagesOrDefault(GetFaults(method))
+                        })
+                        .Where(tuple => tuple.Message != null);
+
+                    foreach (var faultyMethod in allFaultyMethods)
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, faultyMethod.Location,
+                            faultyMethod.Message));
+                    }
+                },
+                SyntaxKind.ClassDeclaration);
+        }
+
+        private static IEnumerable<string> GetFaults(IMethodSymbol methodSymbol)
+        {
+            if (methodSymbol.DeclaredAccessibility != Accessibility.Public)
+            {
+                yield return MakePublicMessage;
+            }
+
+            if (methodSymbol.IsAsync && methodSymbol.ReturnsVoid)
+            {
+                yield return MakeNonAsyncMessage;
+            }
+
+            if (methodSymbol.IsGenericMethod)
+            {
+                yield return MakeNotGenericMessage;
+            }
+        }
+
+        private static string ConcatMessagesOrDefault(IEnumerable<string> faults)
```

Comment:
No memory of such method :) On the other hand, I suppose you could drop `case 2:` because imo it will be handled by the `default`...